src/Client/Request.php

```diff
@@ -138,7 +146,16 @@ public function handleData($data)
         // buffer until double CRLF (or double LF for compatibility with legacy servers)
         if (false !== strpos($this->buffer, "\r\n\r\n") || false !== strpos($this->buffer, "\n\n")) {
             try {
-                list($response, $bodyChunk) = $this->parseResponse($this->buffer);
+                $psrResponse = gPsr\parse_response($this->buffer);
+
+                if ($this->responseIsAnUpgradeResponse($psrResponse)) {
+                    $this->stream->removeListener('data', array($this, 'handleData'));
+
+                    $this->emit('upgrade', array($this->stream, $psrResponse, $this));
```

Comment:
Not sure if this is the right `Request` object to expose, but it's kind of a good idea to expose `a` request object because there could be cases where you want to verify some things about the request **after** upgrade and then close the connection if that verification fails. That's my thought about it. But sure, I might be wrong here.

I actually read a piece of Node docs to see how `upgrade` is being handled by the consumer side. I followed that in adding the upgrade event also. [Here](https://nodejs.org/api/http.html#http_event_upgrade) is a sample piece of how node handles the upgrade.

Sure, If there's a better you think this could be or should be done, I'll be willing to make it work.